File: Exercicios C/Codigos-Livres/PilhaFlexivelInversa.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula{

    int valor;
    struct Celula *proximo;

}Celula;

typedef struct Pilha{

    Celula *topo;

}Pilha;

Pilha* criarPilha(){ // "Construtor"

    Pilha *PilhaFlexivel = (Pilha*)malloc(sizeof(Pilha));

    PilhaFlexivel->topo = NULL;

    return PilhaFlexivel;
}

void push(Pilha *pilha, int valor){

    Celula *celula = (Celula*)malloc(sizeof(Celula));
    celula->valor = valor;
    celula->proximo = pilha->topo;
    pilha->topo = celula;
}

void inverter(Pilha *pilha){
    Celula *atual = NULL;
    Celula *proximo = pilha->topo;
    Celula *anterior = NULL;

    while(atual != NULL){
        proximo = atual->proximo;
        atual->proximo = anterior;
    
        anterior = atual;
        atual = proximo;

    }
}

void mostrar(Pilha *pilha){

    if(pilha->topo == NULL){
        printf("Pilha Vazia");
    }

    Celula *i = pilha->topo;

    while(i != NULL){
        printf("%d ", i->valor);

        i = i->proximo;
    }
}

int main(){
    
    Pilha *pilha = criarPilha();
    int opc;

    do { 
            printf("\n----- MENU -----");
            printf("\nDigite 0 para SAIR");
            printf("\nDigite 1 para INSERIR");
            printf("\nDigite 2 para INVERTER");
            printf("\nDigite 3 para MOSTRAR");
            printf("\nEscolha uma opcao: ");

            scanf("%d", &opc);

            switch(opc) {
                case 1:{
                    printf("\nDigite o valor: ");
                    int valor; 
                    scanf("%d", &valor);
                    push(pilha,valor);
                    break;
            }
                case 2:
                    inverter(pilha);
                    break;
            
                case 3:
                    mostrar(pilha);
                    break;
        }
            
    } while (opc != 0);

    return 0;
    
}

File: Exercicios-JAVA/Codigos-Livres/PilhaFlexivelInversa/PilhaFlexivelInversa.java
import java.util.Scanner;

public class PilhaFlexivelInversa {
    class Celula{
        private int valor;
        private Celula proximo;

        public Celula(int valor){
            this.valor = valor;
            this.proximo = null;
        }

        public int getValor(){
            return valor;
        }
    }

    private Celula topo;

    public PilhaFlexivelInversa(){
        this.topo = null;
    }

    public void push(int valor){
        Celula celula = new Celula(valor);
        celula.proximo = this.topo; 
        this.topo = celula;                              
    }

    public void inverter(){
        Celula anterior = null;
        Celula atual = this.topo;
        Celula proximo = null;

        while(atual != null){
            proximo = atual.proximo;
            atual.proximo = anterior;

            anterior = atual;
            atual = proximo;
        }

        this.topo = anterior;

    }

    public void mostrar(){

        Celula i = this.topo;

        while (i != null) {
            System.out.print(i.getValor() + " ");
            i = i.proximo;
        }

    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        PilhaFlexivelInversa pilha = new PilhaFlexivelInversa();
        int opc;

        do { 
            System.out.println("----- MENU -----");
            System.out.println("Digite 0 para SAIR");
            System.out.println("Digite 1 para INSERIR");
            System.out.println("Digite 2 para INVERTER");
            System.out.println("Digite 3 para MOSTRAR");
            System.out.print("Escolha uma opcao: ");

            opc = scanner.nextInt();

            switch(opc) {
                case 1:{
                    System.out.print("Digite o valor: ");
                    int valor = scanner.nextInt();
                    pilha.push(valor);
                    break;
                }
                case 2:{
                    pilha.inverter();                           
                    break;
                }
                case 3:
                    pilha.mostrar();
                    break;
                default:
                    throw new AssertionError();
            }
            
        } while (opc != 0);

    }
}
